sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala

```diff
@@ -524,14 +524,17 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
     // If this table is cached as an InMemoryRelation, drop the original
     // cached version and make the new version cached lazily.
     val cache = sparkSession.sharedState.cacheManager.lookupCachedData(table)
+
+    // uncache the logical plan.
+    // note this is a no-op for the table itself if it's not cached, but will invalidate all
+    // caches referencing this table.
+    sparkSession.sharedState.cacheManager.uncacheQuery(table, cascade = true)
+
     if (cache.nonEmpty) {
       // save the cache name and cache level for recreation
       val cacheName = cache.get.cachedRepresentation.cacheBuilder.tableName
       val cacheLevel = cache.get.cachedRepresentation.cacheBuilder.storageLevel
 
-      // uncache the logical plan.
-      sparkSession.sharedState.cacheManager.uncacheQuery(table, cascade = true)
-
       // recache with the same name and cache level.
       sparkSession.sharedState.cacheManager.cacheQuery(table, cacheName, cacheLevel)
```

Comment:
Hmm, we will recache the table lazily. But for other caches referencing this table, we just uncache them and don't recache them. Does it sound inconsistent?

Reply:
Nvm, we only recache if the table is cached. So for this case, the table is not cached at all, we don't recache it.

Reply:
Not related to this PR. Since we recache the table lazily, the other caches referencing this table seems to me need to be recached. It is more consistent and seems to me we should not implicitly change the cache status of the other cached query plans. I'd like to recache these other caches if possible. WDYT? @cloud-fan